xds/clients/grpctransport/grpc_transport.go

```diff
@@ -0,0 +1,175 @@
+	v3lrsgrpc "github.com/envoyproxy/go-control-plane/envoy/service/load_stats/v3"
+	v3lrspb "github.com/envoyproxy/go-control-plane/envoy/service/load_stats/v3"
+)
+
+// ServerConfigExtension is an interface to extend the `clients.ServerConfig` for
+// gRPC transport builder. Any implementation needs to implement `ServerConfig`
+// method.
+type ServerConfigExtension interface {
+	ServerConfig() *ServerConfig
+}
+
+// ServerConfig contains the configuration for the gRPC transport server.
+type ServerConfig struct {
+	// Credentials is the credential bundle to be used.
+	Credentials credentials.Bundle
+}
+
+// ServerConfig returns the ServerConfig itself. This method is designed
+// to satisfy interface requirement.
+func (s *ServerConfig) ServerConfig() *ServerConfig {
+	return s
+}
+
+// Builder provides a way to build a gRPC-based transport to an xDS server.
+type Builder struct{}
+
+// Build creates a new gRPC-based transport to an xDS server using the provided
+// options. This involves creating a grpc.ClientConn to the server identified by
+// the server URI in the provided options.
+func (b *Builder) Build(sc clients.ServerConfig) (clients.Transport, error) {
+	if sc.ServerURI == "" {
+		return nil, fmt.Errorf("ServerConfig Uri field in opts cannot be empty")
+	}
+	if sc.Extensions == nil {
+		return nil, fmt.Errorf("ServerConfig Extensions field in opts cannot be nil")
+	}
+	gtsce, ok := sc.Extensions.(ServerConfigExtension)
+	if !ok {
+		return nil, fmt.Errorf("ServerConfig field in opts cannot be anything other than GRPCTransportServerConfigExtension")
+	}
+	gtsc := gtsce.ServerConfig()
+	if gtsc.Credentials == nil {
+		return nil, fmt.Errorf("ServerConfig Credentials field in opts cannot be nil")
+	}
+
+	// Actual implementation of this function will incorporate reference
+	// count map for existing transports and deduplicate transports based on
+	// server URI and credentials. The deduping logic and reference incr/decr
+	// logic is not shown here for brevity.
+	//
+	// Build() increments the count and returns the existing transport. When
+	// Close() is called on the transport, the reference count is decremented.
+	// The transport will be removed from map only when the reference count
+	// reaches zero.
+
+	cc, err := grpc.NewClient(sc.ServerURI, grpc.WithCredentialsBundle(gtsc.Credentials))
+	if err != nil {
+		return nil, fmt.Errorf("error creating grpc client for server uri %s, %v", sc.ServerURI, err)
+	}
+	cc.Connect()
+	return &grpcTransport{cc: cc}, nil
+}
+
+type grpcTransport struct {
+	cc *grpc.ClientConn
+}
+
+func (g *grpcTransport) NewStream(ctx context.Context, method string) (clients.Stream[clients.StreamRequest, any], error) {
+	switch method {
+	case v3adsgrpc.AggregatedDiscoveryService_StreamAggregatedResources_FullMethodName:
+		g.newADSStream(ctx)
+	case v3lrsgrpc.LoadReportingService_StreamLoadStats_FullMethodName:
+		return g.newLRSStream(ctx)
+	}
+	return nil, fmt.Errorf("unsupported method: %v", method)
```

Comment:
@dfawley are you suggesting something like this `stream, err := g.cc.NewStream(ctx, &grpc.StreamDesc{}, methodName)` but then user still needs to pass the method name. So, in case of ads user need to pass "/envoy.service.discovery.v3.AggregatedDiscoveryService/StreamAggregatedResources"

Reply:
oh ok. Looks like grpc.StreamDesc{} need a value for both ads and lrs which is defined in their respective client interface so to differentiate the client we have to enumerate. However, any other implementation can choose to directly create the stream from the passed ctx and methodName